app/models/package_manager/base.rb

```diff
@@ -146,14 +146,20 @@ def self.update(name, sync_version: :all, force_sync_dependencies: false)
       db_project = ensure_project(mapped_project, reformat_repository_url: sync_version == :all)
 
       if self::HAS_VERSIONS
-        if sync_version == :all
-          versions_as_version_objects(raw_project, db_project.name)
-            .each { |v| add_version(db_project, v) }
-            .tap { |vs| remove_missing_versions(db_project, vs) }
-        elsif (version = one_version_as_version_object(raw_project, sync_version))
-          add_version(db_project, version)
-          # TODO: handle deprecation here too
-        end
+        api_versions = if sync_version == :all
+                         versions_as_version_objects(raw_project, db_project.name)
+                       else
+                         [one_version_as_version_object(raw_project, sync_version)]
+                       end
+
+        BulkVersionUpdater.new(
+          db_project: db_project,
+          api_versions: api_versions,
+          repository_source_name: self::HAS_MULTIPLE_REPO_SOURCES ? [self::REPOSITORY_SOURCE_NAME] : nil
+        ).run!
+
+        remove_missing_versions(db_project, api_versions.map(&:version_number)) if sync_version == :all
```

Comment:
Should this use the version numbers from the existing and newly inserted Versions from `BulkVersionUpdater`?

I think this would potentially allow for extra API versions to be kept compared to what we actually found/inserted. Not sure how common that would be in practice, or if that matters

Reply:
that's a great idea, altho could take a bit of code surgery. I'm tempted to introduce the least changes possible in this PR, so I'll just add a note about that for later.